analysisservice/src/main/java/nu/yona/server/analysis/service/AnalysisEngineService.java

```diff
@@ -188,14 +189,22 @@ private void analyzeInsideLock(List<ActivityPayload> payloads, UserAnonymizedDto
 	private void updateLastMonitoredActivityDateIfRelevant(List<ActivityPayload> payloads, UserAnonymizedDto userAnonymized,
 			UserAnonymizedEntityHolder userAnonymizedEntityHolder)
 	{
-		Optional<LocalDate> lastActivityEndTime = payloads.stream().map(payload -> payload.endTime).max(ZonedDateTime::compareTo)
-				.map(ZonedDateTime::toLocalDate);
+		Optional<ActivityPayload> lastActivityPayload = payloads.stream().max((p1, p2) -> p1.endTime.compareTo(p2.endTime));
+		lastActivityPayload
+				.ifPresent(p -> updateLastMonitoredActivityDateIfRelevant(p, userAnonymized, userAnonymizedEntityHolder));
+	}
+
+	private void updateLastMonitoredActivityDateIfRelevant(ActivityPayload lastActivityPayload, UserAnonymizedDto userAnonymized,
+			UserAnonymizedEntityHolder userAnonymizedEntityHolder)
+	{
+		LocalDate lastActivityEndTime = lastActivityPayload.endTime.toLocalDate();
 		Optional<LocalDate> lastMonitoredActivityDate = userAnonymized.getLastMonitoredActivityDate();
-		if (lastActivityEndTime.isPresent()
-				&& lastMonitoredActivityDate.map(d -> d.isBefore(lastActivityEndTime.get())).orElse(true))
+		if (lastMonitoredActivityDate.map(d -> d.isBefore(lastActivityEndTime)).orElse(true))
 		{
-			activityUpdateService.updateLastMonitoredActivityDate(userAnonymizedEntityHolder.getEntity(),
-					lastActivityEndTime.get());
+			DeviceAnonymized deviceAnonymized = deviceService.getDeviceAnonymizedEntity(userAnonymized.getId(),
```

Comment:
Isn't it cheaper to retrieve the `DeviceAnonymized` entity from the `UserAnonymized` entity?

Reply:
The ``userAnonymized`` is a DTO, not an entity, so we cannot get the ``DeviceAnonymized`` entity from it.